### DISCUSSION TITLE
audio_eof_speech inside a class

### DISCUSSION THREAD

**Discussion by iverona** [schreibfaul1/ESP32-audioI2S#349] [Q&A] [answered]
Hi! 

I'm trying to encapsulate a behaviour that includes playing sound in a class, and I'm not able to get audio_eof_speech to work. I've implemented the method as

```
void MyClass::audio_eof_speech(const char *info)
{
    WebSerial.println("I've stopped playing sound");
    is_playing_sound = false;
}
```

but it's (apparently) not being triggered. TTS audio is coming out, this is the only thing that I've not been able to make work. I want to use the flag to stop reading sensors, etc. while audio is being played back to avoid the choppy sound I'm sometimes getting. Any other ideas? I've also thought moving it to a different core, but would like to try this simple approach of using a flag to stop other tasks. 

Thanks!

**Comment by iverona** [comment, marked as answer]
Ah, actually it works by defining `audio_eof_speech` **outside** the class, in the main.cpp... I don't fully understand why it works, I'll do a bit of research.

**Comment by schreibfaul1** [comment]
There is a difference to the usual CallBacks. With CBs, the user usually specifies the name and must ensure that the function exists, otherwise the compiler will complain.
I chose the 'Weak Attribute'. So the compiler doesn't care if the function exists or not.

**Reply by iverona** [reply]
Thank you! My question was more in the sense of where to define the extern weak function, as I'd like to have it all inside the class where the Audio object is defined, but I've only been able to make it work defining the event functions in the main file. I guess the compiler only searches the generic namespace, and not inside class member functions. I'm far away from being a C++ expert. 

Finally I had to implement something blocking, like discussed [here](https://github.com/schreibfaul1/ESP32-audioI2S/discussions/287), and now the audio output (both TTS and mp3) is ok 90% of the time. Any other ideas before moving it all to a different ESP32 core? 

Thanks for this great library schreibfaul1.
